### PR TITLE
KAFKA-17639 Add Java 23 to CI build matrix

### PR DESCRIPTION
**Pull request by m1a2st**
Jira: https://issues.apache.org/jira/browse/KAFKA-17639
Add support java 23 and add java version 23 on ci

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
@m1a2st I add WIP to the title since this PR can't be merged until we upgrade spotless check.

File: jmh-benchmarks/src/main/java/org/apache/kafka/jmh/util/ByteUtilsBenchmark.java [hunk 1]
```diff
@@ -41,17 +41,17 @@
...
  * This benchmark calculates the empirical evidence of different implementation for encoding/decoding a protobuf
  * <a href="https://protobuf.dev/programming-guides/encoding/#varints">VarInt</a> and VarLong.
  *
  * The benchmark uses JMH and calculates results for different sizes of variable length integer. We expect most of the
  * usage in Kafka code base to be 1 or 2 byte integers.
  */
+@OutputTimeUnit(TimeUnit.SECONDS)
+@Fork(3)
+@Warmup(iterations = 3, time = 1)
+@Measurement(iterations = 5, time = 1)
```
Comment (m1a2st) [inline jmh-benchmarks/src/main/java/org/apache/kafka/jmh/util/ByteUtilsBenchmark.java:54]: This change is used to resolve the `[dangling-doc-comments] documentation comment is not attached to any declaration` error

File: gradle/dependencies.gradle [hunk 2]
```diff
@@ -47,23 +47,14 @@ if ( !versions.scala.contains('-') ) {
 }
 
 // mockito >= 5.5 is required for Java 21 and mockito 5.x requires at least Java 11
-// mockito 4.11 is used with Java 8 and Scala 2.13
-String mockitoVersion
-if (JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_11))
-  mockitoVersion = "5.10.0"
-else
-  mockitoVersion = "4.11.0"
+String mockitoVersion = "5.14.1"
```
Comment (m1a2st) [inline gradle/dependencies.gradle]: in my local machine, it has error about 
```
Caused by: java.lang.IllegalArgumentException: Java 23 (67) is not supported by the current version of Byte Buddy which officially supports Java 22 (66) - update Byte Buddy or set net.bytebuddy.experimental as a VM property
```
And I find mockito is resolve this problem in this [issue](https://github.com/mockito/mockito/pull/3459), thus I upgrade the mockito version.

File: gradle/dependencies.gradle [hunk 3]
```diff
@@ -47,23 +47,14 @@ if ( !versions.scala.contains('-') ) {
 }
 
 // mockito >= 5.5 is required for Java 21 and mockito 5.x requires at least Java 11
-// mockito 4.11 is used with Java 8 and Scala 2.13
-String mockitoVersion
-if (JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_11))
-  mockitoVersion = "5.10.0"
-else
-  mockitoVersion = "4.11.0"
+String mockitoVersion = "5.14.2"
```
Comment (chia7712) [inline gradle/dependencies.gradle]: Could you please add `mockitoVersion` to `versions`?

File: gradle/dependencies.gradle [hunk 4]
```diff
@@ -47,23 +47,14 @@ if ( !versions.scala.contains('-') ) {
...
 // mockito 5.x has two mocking strategies: inline (mockito-core) and subclass (mockito-subclass).
 // The default strategy (i.e. what `mockito-core` uses) changed to `inline` in 5.x because it works better with newer
 // Java versions.
 // We always use the `inline` strategy.
-String mockitoArtifactName
-if (mockitoVersion.startsWith("4."))
-  mockitoArtifactName = "mockito-inline"
-else
-  mockitoArtifactName = "mockito-core"
+String mockitoArtifactName = "mockito-core"
```
Comment (chia7712) [inline gradle/dependencies.gradle]: we don't need this `mockitoArtifactName`, right?

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for this patch!

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -104,7 +104,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        java: [ 21, 11 ]  # If we change these, make sure to adjust ci-complete.yml
+        java: [ 23, 21, 11 ]  # If we change these, make sure to adjust ci-complete.yml
```
Comment (chia7712) [inline .github/workflows/build.yml]: Could you please remove 21 from tests? two suits are good enough I think

**Comment by chia7712** [comment]
@m1a2st please check https://bugs.openjdk.org/browse/JDK-8331987 for the behavior change on `CompletableFuture`. I think we should check the `ExecutionException` and `CancellationException` in get methods. WDYT?

**Comment by chia7712** [comment]
Additionally, `SecurityManager` is unsupported by java 23 (see https://cwiki.apache.org/confluence/display/KAFKA/KIP-1006%3A+Remove+SecurityManager+Support), so you have to update `SaslPlainSslEndToEndAuthorizationTest` to use `SecurityManagerCompatibility.get().current()`

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 6]
```diff
@@ -180,7 +180,7 @@ public T get(long timeout, TimeUnit unit) throws InterruptedException, Execution
             return completableFuture.get(timeout, unit);
         } catch (ExecutionException e) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java]: we should unwrap the `CancellationException` for java 23.
```java
    public T get() throws InterruptedException, ExecutionException {
        try {
            return completableFuture.get();
        } catch (ExecutionException | CancellationException e) {
            maybeThrowCancellationException(e.getCause());
            throw e;
        }
    }
```

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 7]
```diff
@@ -252,6 +252,9 @@ public String toString() {
         } catch (Exception e) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:276]: we should handle `CancellationException` first
```java
        try {
            value = completableFuture.getNow(null);
        } catch (CancellationException e) {
            if (e.getCause() instanceof CancellationException) {
                exception = e.getCause();
            } else exception = e;
        } catch (CompletionException e) {
            exception = e.getCause();
        } catch (Exception e) {
            exception = e;
        }
```

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 8]
```diff
@@ -165,7 +165,7 @@ public T get() throws InterruptedException, ExecutionException {
             return completableFuture.get();
         } catch (ExecutionException e) {
             maybeThrowCancellationException(e.getCause());
-            throw e;
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:168]: ditto
```java
        try {
            return completableFuture.get();
        } catch (ExecutionException | CancellationException e) {
            maybeThrowCancellationException(e.getCause());
            throw e;
        }
```

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for this patch. I leave some sample code for you. PTAL

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 9]
```diff
@@ -192,6 +192,14 @@ public T get(long timeout, TimeUnit unit) throws InterruptedException, Execution
     public T getNow(T valueIfAbsent) throws ExecutionException {
         try {
             return completableFuture.getNow(valueIfAbsent);
+        } catch (CancellationException e) {
+            // In Java 23, When a CompletableFuture is cancelled, getNow() will throw a CompletionException wrapping a 
+            // CancellationException. whereas in Java < 23, it throws a CompletionException directly.
+            if (e.getCause() instanceof CancellationException) {
+                throw (CancellationException) e.getCause();
+            } else {
+                throw new CancellationException(e.getMessage());
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java]: why not throwing `e.getCause` directly?

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 10]
```diff
@@ -178,7 +178,7 @@ public T get(long timeout, TimeUnit unit) throws InterruptedException, Execution
             TimeoutException {
         try {
             return completableFuture.get(timeout, unit);
-        } catch (ExecutionException e) {
+        } catch (ExecutionException | CancellationException e) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:187]: please add comment

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 11]
```diff
@@ -163,7 +163,7 @@ private void maybeThrowCancellationException(Throwable cause) {
     public T get() throws InterruptedException, ExecutionException {
         try {
             return completableFuture.get();
-        } catch (ExecutionException e) {
+        } catch (ExecutionException | CancellationException e) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:169]: please add comment

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for this patch

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java (hunk 9 above)
Comment (m1a2st) [reply clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java]: The method exception signature doesn't have `Throwable`, so we can't  throw `e.getCause()`.
but we can throw `CancellationException e` directory.

File: clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java [hunk 12]
```diff
@@ -163,7 +163,9 @@ private void maybeThrowCancellationException(Throwable cause) {
     public T get() throws InterruptedException, ExecutionException {
         try {
             return completableFuture.get();
-        } catch (ExecutionException e) {
+            // In Java 23, When a CompletableFuture is cancelled, get() will throw a CancellationException wrapping a
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:166]: Please add the JDK link to the comment

**Review by chia7712 — APPROVED** [review]

**Comment by apoorvmittal10** [comment]
@chia7712 @m1a2st I was wondering if we should also unwrap CancellationException in Errors.java: https://github.com/apache/kafka/pull/17821. Though I think we should regardless of JDK bug. Wdyt?

**Comment by apoorvmittal10** [comment]
> @chia7712 @m1a2st I was wondering if we should also unwrap CancellationException in Errors.java: #17821. Though I think we should regardless of JDK bug. Wdyt?

My bad, not needed.

**Pull request merged**